File: Source/NSExpressionDescription+CoreDataKit.m

#import "NSExpressionDescription+CoreDataKit.h"


@implementation NSExpressionDescription (CoreDataKit)

- (instancetype)initWithFunction:(NSString *)function key:(NSString *)key resultType:(NSAttributeType)type {
	NSExpression *keyExpression = [NSExpression expressionForKeyPath:key];
	
	typeof(self) this = [self initWithResultType:type];
	this.expression = [NSExpression expressionForFunction:function arguments:@[keyExpression]];
	this.name = key;
	return this;
}

- (instancetype)initWithResultType:(NSAttributeType)type {
	typeof(self) this = [self init];
	this.expressionResultType = type;
	
	if (type == NSObjectIDAttributeType) {
		this.expression = [NSExpression expressionForEvaluatedObject];
		this.name = @"objectID";
	}
	
	return this;
}

@end

File: Source/NSExpressionDescription+CoreDataKit.swift
//
//  NSExpressionDescription.swift
//  CoreDataKit
//
//  Created by Felix Gabel on 21/07/14.
//  Copyright (c) 2014 Felix Gabel. All rights reserved.
//

import CoreData


public extension NSExpressionDescription {

	public class func descriptionWithFunction(function:String, key:String, resultType:NSAttributeType) -> NSExpressionDescription {
		
		let keyExpression = NSExpression(forKeyPath:key)
		let expression = NSExpression(forFunction:function, arguments:[keyExpression])
		
		let description = NSExpressionDescription()
		description.expressionResultType = resultType
		description.expression = expression
		description.name = key
		
		return description
	}
	
	public class func objectIDDescription() -> NSExpressionDescription {
		
		let description = NSExpressionDescription()
		description.expressionResultType = .ObjectIDAttributeType
		description.expression = NSExpression.expressionForEvaluatedObject()
		description.name = "objectID"
		
		return description
	}
}
